intermediate_source/rpc_param_server_tutorial.rst

```diff
@@ -0,0 +1,398 @@
+           self.num_gpus = num_gpus
+           device = torch.device(
+               "cuda:0" if torch.cuda.is_available() and self.num_gpus > 0 else "cpu")
+           print(f"Putting first 2 convs on {str(device)}")
+           # Put conv layers on the first cuda device, or CPU if no cuda device
+           self.conv1 = nn.Conv2d(1, 32, 3, 1).to(device)
+           self.conv2 = nn.Conv2d(32, 64, 3, 1).to(device)
+           # Put rest of the network on the 2nd cuda device, if there is one
+           if "cuda" in str(device) and num_gpus > 1:
+               device = torch.device("cuda:1")
+
+           print(f"Putting rest of layers on {str(device)}")
+           self.dropout1 = nn.Dropout2d(0.25).to(device)
+           self.dropout2 = nn.Dropout2d(0.5).to(device)
+           self.fc1 = nn.Linear(9216, 128).to(device)
+           self.fc2 = nn.Linear(128, 10).to(device)
+
+       def forward(self, x):
+           x = self.conv1(x)
+           x = F.relu(x)
+           x = self.conv2(x)
+           x = F.max_pool2d(x, 2)
+
+           x = self.dropout1(x)
+           x = torch.flatten(x, 1)
+           # Move tensor to next device if necessary
+           next_device = next(self.fc1.parameters()).device
+           x = x.to(next_device)
+
+           x = self.fc1(x)
+           x = F.relu(x)
+           x = self.dropout2(x)
+           x = self.fc2(x)
+           output = F.log_softmax(x, dim=1)
+           return output
+
+Next, let's define some helper functions that will be useful for the rest of our script. The following uses `rpc_sync <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.rpc_sync>`_ and `RRef <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.RRef>`_ in order to define a function that invokes a given method on an object living on a remote node. Below, our handle to the remote object is given by the ``rref`` argument, and we run it on its owning node: ``rref.owner()``. On the caller node, we run this command synchronously through the use of ``rpc_sync``\ , meaning that we will block until a response is received.
+
+.. code-block:: python
+
+   # --------- Helper Methods --------------------
+
+   # On the local node, call a method with first arg as the value held by the
+   # RRef. Other args are passed in as arguments to the function called.
+   # Useful for calling instance methods. method could be any matching function, including 
+   # class methods.
+   def call_method(method, rref, *args, **kwargs):
+       return method(rref.local_value(), *args, **kwargs)
+
+   # Given an RRef, return the result of calling the passed in method on the value
+   # held by the RRef. This call is done on the remote node that owns
+   # the RRef and passes along the given argument.
+   # Example: If the value held by the RRef is of type Foo, then
+   # remote_method(Foo.bar, rref, arg1, arg2) is equivalent to calling
+   # <foo_instance>.bar(arg1, arg2) on the remote node and getting the result
+   # back.
+
```

Comment:
remove empty line?

Reply:
Shoulnd' there be a space here since we're going to a different method? `def remote_method` begins on the next line.